File: solutions/019-remove-nth-node-from-end-of-list/solution.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func sliceToList(nums []int) *ListNode {
	dummy := &ListNode{}
	cur := dummy
	for _, v := range nums {
		cur.Next = &ListNode{Val: v}
		cur = cur.Next
	}
	return dummy.Next
}

func listToSlice(head *ListNode) []int {
	out := []int{}
	for head != nil {
		out = append(out, head.Val)
		head = head.Next
	}
	return out
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummy := &ListNode{0, head}
	first := dummy
	second := dummy
	for i := 0; i < n+1; i++ {
		first = first.Next
	}
	for first != nil {
		first = first.Next
		second = second.Next
	}
	second.Next = second.Next.Next
	return dummy.Next

}

func main() {
	head := sliceToList([]int{1, 2, 3, 4, 5})
	got := listToSlice(removeNthFromEnd(head, 2))
	want := []int{1, 2, 3, 5}
	status := "PASS"
	if len(got) != len(want) {
		status = "FAIL"
	}
	fmt.Printf("%s | %v (expected %v)\n", status, got, want)
}

File: solutions/019-remove-nth-node-from-end-of-list/solution.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def remove_nth_from_end(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        first = dummy
        second = dummy
        for _ in range(n + 1):
            first = first.next
        while first is not None:
            first = first.next
            second = second.next
        second.next = second.next.next
        return dummy.next


if __name__ == "__main__":
    import sys
    from pathlib import Path
    sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "_helpers"))
    from testutil import build_list, list_to_slice
    sol = Solution()
    head = build_list([1, 2, 3, 4, 5])
    got = list_to_slice(sol.remove_nth_from_end(head, 2))
    status = "PASS" if got == [1, 2, 3, 5] else "FAIL"
    print(f"{status} | => {got} (expected [1, 2, 3, 5])")
